# Intro to Artificial Intelligence with Python

## Part I - Search

Harvard CS50 Introduction to Artificial Intelligence with Python is an online course that I took in the Spring of 2020. It consisted of 6 lectures of which I have a notebook for each. Each lecture had 2 projects, those are located in the projects folder in the same directory as this notebook.

[Course Link](https://cs50.harvard.edu/ai/)

[Lecture Link](https://www.youtube.com/watch?v=D5aJNFWsWew&list=PLhQjrBD2T382Nz7z1AEXmioc27axa19Kv&index=2)

---
## Basic Search AI Definitions

* **agent** - entity that perceives its environment and actus upon that environemnt


* **state** - a configuration of the agent and its environment


* **initial state** - state where the agent begins, starting point for an algorithm


* **actions** - choices that can be made in any given state, example possible moves on a game board given current state or game situation
    * function example: Actions(state): returns the set of actions that can be executed in state s


* **transition model** - a description of what state results from performing any applicable action in any state
    * function example: Results(s,a): returns the state resulting from perfoming action a in state s


* **state space** - the set of all states reachable from the initial state by any sequence of actions 


* **goal test** - way to determine whether a given state is a goal state


* **path cost** - numerical cost associated with a given path


* **solution** - a sequence of actions that leads from the initial state to a goal state


* **optimal solution** - a solution that has the lowest path cost among all solutions


## Search Problem General Steps:
* initial state
* actions
* transition model
* goal test
* path cost function

## Node
A data structure that keeps track of (specifically for search problems): 
* a state
* a parent node (node that genreated this node)
* an action (action applied to parent to get node)
* a path cost (from initial state to node)

## Approach to Solving Problems
* Start with a frontier (all possible states available at a given time) that contains just the initial state
* Repeat:
    * If the frontier is empty, no solution
    * Remove a node from the frontier
    * If node contains goal state, return the solution
    * Exapand node, add resulting nodes to the frontier
    * then repeat the above process until either solution or no solution
    
The above approach works for single directional action movements (that is only one way, not going back), but if an action can be bi-directional, then this approach will fail as infinite loops are possible. To fix this approach, state has to be stored in some manner.

## Revised Approach
* Start with a frontier (all possible states available at a given time) that contains just the initial state
* Start with an exmpty explored set
* Repeat:
    * If the frontier is empty, no solution
    * Remove a node from the frontier
    * If node contains goal state, return the solution
    * Add node to explored set
    * Exapand node, add resulting nodes to the frontier if they aren't already in the frontier OR in the explored set.

---
## Types of Search Algorithms
There are two basic types of search algorithms:
1. **uninformed** - strategy that uses no problem-specific knowledge
2. **informed**   - strategies that use problem-specific knowledge to find solutions more effeciently

---

## Uninformed Search Algorithms

### Depth First Search (DFS)
A search algorithim that always expands the deepest node in a frontier
* stack based, checks nodes last-in-first-out (LIFO)

### Breadth First Search (BFS)
A search algorithm that always expands the shallowest node in a frontier
* queue based, checks nodes first-in-first-out (FIFO)

---

## Informed Search  Algorithms

### Greedy Best First Search (Manhatten Distance)
A search algorithm that expands the node that is closest to the goal, as estimated by a heuristic function h(n)

### A* Search
A search algorithm that expands node with lowest value of g(n) + h(n) where:
* g(n) = cost to reach node (how many steps needed to get to the current node)
* h(n) = estimated cost to goal (distance from node to goal)

This algorithm can be optimal IF:
* h(n) is admissible (never overestimates the true cost), and
* h(n) is consistent (for every node n  where current h(n) + c is < next h(ni) + c)

---

## Adversarial Search Algorithms
Search algorithms where there is an adversarial element introduced that attempts to block or prevent the search algorithm from completion. Most often found in games, like tic-tac-toe

### Minimax (works well with 2-player games, tic-tac-toe, checkers, chess)
This is a search algorithm that assigns ranked values to game situations. For example, the game of tic-tac-toe has three possible outcomes:

1. x-wins, assign 1 as the outcome, x is the MAX player
2. tie, assign 0 as the outcome
3. o-wins, assign -1 as the outcome, o i sthe MIN player

In this case, the x-player will choose actions that can maximize their score,
whereas the o-player will choose actions that can minimize their score, both will 
choose a tie over letting the other win. 

Example of steps in the Minimax process with tic-tac-toe:
* So : initial state
* function PLAYER(s)   : returns which player to move in state s
* function ACTIONS(s)  : returns legal moves in state s
* function RESULT(s,a) : returns state after action a taken in state s
* function TERMINAL(s) : checks if state s is a terminal (game over) state
* function UTILITY(s)  : final numerical value for terminal state s, 1, 0, or -1

Detailed steps:
* Given a state s:
    * MAX pics action a in ACTIONS(s) that produces the highest value of MIN-VALUE(RESULTS(s,a)) out of all possible MIN player moves. 
    * MIN picks picks a in ACTIONS(s) that produces the smallest value of MAX-VALUE(RESULTS(s,a)) out of all possible MAX player moves. 
    
A sample Max-Value function could look like:

In [ ]:
'''
# MAX Player Logic:
def MAX-VALUE(s):
    if TERMINAL(s):
        return UTILITY(s)
    else:
        v = -infinity or some really low number -9999999999999999
        for action in ACTIONS(state):
            # gets the largest min value of all possible future MIN player moves.
            v = MAX(v, MIN-VALUE(RESULT(state, action)))
        return v
        
# MIN Player Logic:
def MIN-VALUE(s):
    if TERMINAL(s):
        return UTILITY(s)
    else:
        v = +infinity or some really large number +9999999999999999
        for action in ACTIONS(state):
            # gets the smallest max value of all possible future MAX player moves.
            v = MIN(v, MAX-VALUE(RESULT(state, action)))
        return v

'''

## Adversarial Search Algorithms Continued...

The above psuedo-code works for small games, but as games get more complex this methodology will be slow due to the recusive nature of the algorithm. There are methods for optimizing the Minimax algorithm for better efficiency such as:

### Alpha - Beta Pruning
Alpha–beta pruning is a search algorithm that seeks to decrease the number of nodes that are evaluated by the minimax algorithm in its search tree. It is an adversarial search algorithm used commonly for machine playing of two-player games.

### Depth Limited Minimax
Looks at only a certain level of possible future actions rather than all to save computational power. 

* Evaluation Function - estimates the expected utility of the game from a given state

An example of this would be with chess, say a white player posses a 0.80% chance of winning for a certain action, the evaluation function would evaluate the other actions to see if they are better or worse and select the one with the highest outcome. The creation of the estimations and their quality is the difficult part. 

## Maze Examples with Depth First Search and Breadth First Search

In [27]:
# %load maze.py
import sys

class Node():
    def __init__(self, state, parent, action):
        self.state = state
        self.parent = parent
        self.action = action

'''
Depth First Search, Last-In-First-Out (LIFO)
'''
class StackFrontier():
    def __init__(self):
        self.frontier = []

    def add(self, node):
        self.frontier.append(node)
    
    # Checks for particular state
    def contains_state(self, state):
        return any(node.state == state for node in self.frontier)

    def empty(self):
        return len(self.frontier) == 0
    
    # Removes last node from frontier
    def remove(self):
        if self.empty():
            raise Exception("empty frontier")
        else:
            node = self.frontier[-1]
            self.frontier = self.frontier[:-1]
            return node

'''
Breadth First Search, First-In-First_Out (FIFO)
'''
class QueueFrontier(StackFrontier):
    # Removes first node from frontier
    def remove(self):
        if self.empty():
            raise Exception("empty frontier")
        else:
            node = self.frontier[0]
            self.frontier = self.frontier[1:]
            return node

'''
This class takes in a maze from a text file and 
Solves it using either DFS or BFSSearch algorithms
as selected in the solve function within the class
'''
class Maze():
    def __init__(self, filename):

        # Read file and set height and width of maze
        with open(filename) as f:
            contents = f.read()

        # Validate start and goal
        if contents.count("A") != 1:
            raise Exception("maze must have exactly one start point")
        if contents.count("B") != 1:
            raise Exception("maze must have exactly one goal")

        # Determine height and width of maze
        contents = contents.splitlines()
        self.height = len(contents)
        self.width = max(len(line) for line in contents)

        # Keep track of walls
        self.walls = []
        for i in range(self.height):
            row = []
            for j in range(self.width):
                try:
                    # Create start state
                    if contents[i][j] == "A":
                        self.start = (i, j)
                        row.append(False)
                    # Create goal state
                    elif contents[i][j] == "B":
                        self.goal = (i, j)
                        row.append(False)
                    elif contents[i][j] == " ":
                        row.append(False)
                    else:
                        row.append(True)
                except IndexError:
                    row.append(False)
            self.walls.append(row)

        self.solution = None


    def print(self):
        solution = self.solution[1] if self.solution is not None else None
        print()
        for i, row in enumerate(self.walls):
            for j, col in enumerate(row):
                if col:
                    print("█", end="")
                elif (i, j) == self.start:
                    print("A", end="")
                elif (i, j) == self.goal:
                    print("B", end="")
                elif solution is not None and (i, j) in solution:
                    print("*", end="")
                else:
                    print(" ", end="")
            print()
        print()


    def neighbors(self, state):
        row, col = state
        candidates = [
            ("up", (row - 1, col)),
            ("down", (row + 1, col)),
            ("left", (row, col - 1)),
            ("right", (row, col + 1))
        ]

        result = []
        for action, (r, c) in candidates:
            if 0 <= r < self.height and 0 <= c < self.width and not self.walls[r][c]:
                result.append((action, (r, c)))
        return result


    def solve(self):
        """Finds a solution to maze, if one exists."""

        # Keep track of number of states explored
        self.num_explored = 0

        # Step 1: 
        # Initialize frontier to initial starting position
        start = Node(state=self.start, parent=None, action=None)
        frontier = StackFrontier()
        #frontier = QueueFrontier()
        frontier.add(start)
        
        print('yyyyyyyy')
        print(frontier.frontier[0].state)
        print('dkfljsdkfj')
        
        # Step 2: 
        # Start with an empty explored set
        self.explored = set()

        # Step 3: 
        # Repeat Until Solution Found
        while True:

            # Step 3a: 
            # If nothing left in frontier, no solution
            if frontier.empty():
                raise Exception("no solution")

            # Step 3b:
            # Choose a node from the frontier
            node = frontier.remove()
            self.num_explored += 1

            # Step 3c: 
            # Check if node from 3b is goal, if so, solution found
            # this section explained at 42 min in lecture
            if node.state == self.goal:
                actions = []
                cells = []
                while node.parent is not None:
                    actions.append(node.action)
                    cells.append(node.state)
                    node = node.parent
                actions.reverse()
                cells.reverse()
                self.solution = (actions, cells)
                return

            # Step 3d:
            # If 3c fails, add node from 3b to explored set
            self.explored.add(node.state)

            print(self.explored)
            
            # Step 3e. 
            # Expand node, add resulting (neighbor) nodes 
            # to the frontier if they aren't already there 
            # OR if they are not already in the explored set.
            print(node.state)
            for action, state in self.neighbors(node.state):
                if not frontier.contains_state(state) and state not in self.explored:
                    child = Node(state=state, parent=node, action=action)
                    frontier.add(child)

    def output_image(self, filename, show_solution=True, show_explored=False):
        from PIL import Image, ImageDraw
        cell_size = 50
        cell_border = 2

        # Create a blank canvas
        img = Image.new(
            "RGBA",
            (self.width * cell_size, self.height * cell_size),
            "black"
        )
        draw = ImageDraw.Draw(img)

        solution = self.solution[1] if self.solution is not None else None
        for i, row in enumerate(self.walls):
            for j, col in enumerate(row):

                # Walls
                if col:
                    fill = (40, 40, 40)

                # Start
                elif (i, j) == self.start:
                    fill = (255, 0, 0)

                # Goal
                elif (i, j) == self.goal:
                    fill = (0, 171, 28)

                # Solution
                elif solution is not None and show_solution and (i, j) in solution:
                    fill = (220, 235, 113)

                # Explored
                elif solution is not None and show_explored and (i, j) in self.explored:
                    fill = (212, 97, 85)

                # Empty cell
                else:
                    fill = (237, 240, 252)

                # Draw cell
                draw.rectangle(
                    ([(j * cell_size + cell_border, i * cell_size + cell_border),
                      ((j + 1) * cell_size - cell_border, (i + 1) * cell_size - cell_border)]),
                    fill=fill
                )

        img.save(filename)

In [28]:
test = Maze('data/maze1.txt')
test.solve()

yyyyyyyy
(5, 0)
dkfljsdkfj
{(5, 0)}
(5, 0)
{(5, 0), (4, 0)}
(4, 0)
{(4, 1), (5, 0), (4, 0)}
(4, 1)
{(4, 2), (4, 1), (5, 0), (4, 0)}
(4, 2)
{(4, 3), (5, 0), (4, 2), (4, 1), (4, 0)}
(4, 3)
{(4, 4), (4, 3), (5, 0), (4, 2), (4, 1), (4, 0)}
(4, 4)
{(4, 4), (4, 3), (5, 0), (3, 4), (4, 2), (4, 1), (4, 0)}
(3, 4)
{(4, 4), (4, 3), (5, 0), (3, 4), (4, 2), (4, 1), (2, 4), (4, 0)}
(2, 4)
{(4, 4), (4, 3), (5, 0), (3, 4), (4, 2), (2, 5), (4, 1), (2, 4), (4, 0)}
(2, 5)
{(4, 4), (1, 5), (4, 3), (5, 0), (3, 4), (4, 2), (2, 5), (4, 1), (2, 4), (4, 0)}
(1, 5)


---
# Mazes

## Maze 1 (DFS)

In [2]:
m = Maze('data/maze1.txt')
print("Maze 1:")
m.print()
print("Solving...")
m.solve()
print("States Explored:", m.num_explored)
print("Solution:")
m.print()
m.output_image("data/maze1_dfs.png", show_explored=True)

Maze 1:

█████B█
█████ █
████  █
████ ██
     ██
A██████

Solving...
States Explored: 11
Solution:

█████B█
█████*█
████**█
████*██
*****██
A██████



<img src='data/maze1_dfs.png'>

## Maze 2 (DFS)

In [3]:
m = Maze('data/maze2.txt')
print("Maze 2:")
m.print()
print("Solving...")
m.solve()
print("States Explored:", m.num_explored)
print("Solution:")
m.print()
m.output_image("data/maze2_dfs.png", show_explored=True)

Maze 2:

███                 █████████
█   ███████████████████   █ █
█ ████                █ █ █ █
█ ███████████████████ █ █ █ █
█                     █ █ █ █
█████████████████████ █ █ █ █
█   ██                █ █ █ █
█ █ ██ ███ ██ █████████ █ █ █
█ █    █   ██B█         █ █ █
█ █ ██ ████████████████ █ █ █
███ ██             ████ █ █ █
███ ██████████████ ██ █ █ █ █
███             ██    █ █ █ █
██████ ████████ ███████ █ █ █
██████ ████             █   █
A      ██████████████████████

Solving...
States Explored: 194
Solution:

███                 █████████
█   ███████████████████   █ █
█ ████                █ █ █ █
█ ███████████████████ █ █ █ █
█                     █ █ █ █
█████████████████████ █ █ █ █
█   ██********        █ █ █ █
█ █ ██*███ ██*█████████ █ █ █
█ █****█   ██B█         █ █ █
█ █*██ ████████████████ █ █ █
███*██             ████ █ █ █
███*██████████████ ██ █ █ █ █
███****         ██    █ █ █ █
██████*████████ ███████ █ █ █
██████*████             █   █
A******██████████

<img src='data/maze2_dfs.png'>

**Note all the red blocks above show how many states were explored to get from initial state to goal, total of 194 steps**

## Maze 2 (BFS)

In [6]:
m = Maze('data/maze2.txt')
print("Maze 2:")
m.print()
print("Solving...")
m.solve()
print("States Explored:", m.num_explored)
print("Solution:")
m.print()
m.output_image("data/maze2_bfs.png", show_explored=True)

Maze 2:

███                 █████████
█   ███████████████████   █ █
█ ████                █ █ █ █
█ ███████████████████ █ █ █ █
█                     █ █ █ █
█████████████████████ █ █ █ █
█   ██                █ █ █ █
█ █ ██ ███ ██ █████████ █ █ █
█ █    █   ██B█         █ █ █
█ █ ██ ████████████████ █ █ █
███ ██             ████ █ █ █
███ ██████████████ ██ █ █ █ █
███             ██    █ █ █ █
██████ ████████ ███████ █ █ █
██████ ████             █   █
A      ██████████████████████

Solving...
States Explored: 77
Solution:

███                 █████████
█   ███████████████████   █ █
█ ████                █ █ █ █
█ ███████████████████ █ █ █ █
█                     █ █ █ █
█████████████████████ █ █ █ █
█   ██********        █ █ █ █
█ █ ██*███ ██*█████████ █ █ █
█ █****█   ██B█         █ █ █
█ █*██ ████████████████ █ █ █
███*██             ████ █ █ █
███*██████████████ ██ █ █ █ █
███****         ██    █ █ █ █
██████*████████ ███████ █ █ █
██████*████             █   █
A******███████████

<img src='data/maze2_bfs.png'>

**The BFS algorithm only took 77 steps to find the solution**

## Maze 3 (DFS)

In [10]:
m = Maze('data/maze3.txt')
print("Maze 3:")
m.print()
print("Solving...")
m.solve()
print("States Explored:", m.num_explored)
print("Solution:")
m.print()
m.output_image("data/maze3_dfs.png", show_explored=True)

Maze 3:

██    █
██ ██ █
█B █  █
█ ██ ██
     ██
A██████

Solving...
States Explored: 17
Solution:

██****█
██*██*█
█B*█**█
█ ██*██
*****██
A██████



<img src='data/maze3_dfs.png'>

## Maze 3 (BFS)

In [7]:
m = Maze('data/maze3.txt')
print("Maze 3:")
m.print()
print("Solving...")
m.solve()
print("States Explored:", m.num_explored)
print("Solution:")
m.print()
m.output_image("data/maze3_bfs.png", show_explored=True)

Maze 3:

██    █
██ ██ █
█B █  █
█ ██ ██
     ██
A██████

Solving...
States Explored: 6
Solution:

██    █
██ ██ █
█B █  █
█*██ ██
**   ██
A██████



<img src='data/maze3_bfs.png'>

**For maze 3 DFS did not find the optimal solution, but BFS did**